pkg/encoding/kzgEncoder/multiframe.go

```diff
@@ -0,0 +1,212 @@
+	n := len(samples)
+	commits := make([]bls.G1Point, m)
+	D := params.ChunkLen
+
+	var tmp bls.Fr
+
+	// first term
+	// get coeffs to compute the aggregated commitment
+	// note the coeff is affected by how many chunks are validated per blob
+	// if x chunks are sampled from one blob, we need to compute the sum of all x random field element corresponding to each sample
+	aggCommitCoeffs := make([]bls.Fr, m)
+	for k := 0; k < n; k++ {
+		s := samples[k]
+		row := s.RowIndex
+		bls.AddModFr(&aggCommitCoeffs[row], &aggCommitCoeffs[row], &randomsFr[k])
+		bls.CopyG1(&commits[row], &s.Commitment)
+	}
+
+	aggCommit := bls.LinCombG1(commits, aggCommitCoeffs)
+
+	// second term
+	// compute the aggregated interpolation polynomial
+	aggPolyCoeffs := make([]bls.Fr, D)
+
+	// we sum over the weighted coefficients (by the random field element) over all D monomial in all n samples
+	for k := 0; k < n; k++ {
+		coeffs := samples[k].Coeffs
+
+		rk := randomsFr[k]
+		// for each monomial in a given polynomial, multiply its coefficient with the corresponding random field,
+		// then sum it with others. Given ChunkLen (D) is identical for all samples in a subBatch.
+		// The operation is always valid.
+		for j := uint64(0); j < D; j++ {
+			bls.MulModFr(&tmp, &coeffs[j], &rk)
+			bls.AddModFr(&aggPolyCoeffs[j], &aggPolyCoeffs[j], &tmp)
+		}
+	}
+
+	// All samples in a subBatch has identical chunkLen
+	aggPolyG1 := bls.LinCombG1(ks.Srs.G1[:D], aggPolyCoeffs)
+
+	// third term
+	// leading coset is an evaluation index, here we compute the weighted leading coset evaluation by random fields
+	lcCoeffs := make([]bls.Fr, n)
+
+	// get leading coset powers
+	leadingDs := make([]bls.Fr, n)
+
+	for k := 0; k < n; k++ {
+
+		// got the leading coset field element
+		h := ks.ExpandedRootsOfUnity[samples[k].X]
+		var hPow bls.Fr
+		bls.CopyFr(&hPow, &bls.ONE)
+
+		// raising the power for each leading coset
+		for j := uint64(0); j < D; j++ {
+			bls.MulModFr(&tmp, &hPow, &h)
+			bls.CopyFr(&hPow, &tmp)
+		}
+		bls.CopyFr(&leadingDs[k], &hPow)
+	}
+
+	// applying the random weights to leading coset elements
+	for k := 0; k < n; k++ {
+		rk := randomsFr[k]
+		bls.MulModFr(&lcCoeffs[k], &rk, &leadingDs[k])
+	}
+
+	offsetG1 := bls.LinCombG1(proofs, lcCoeffs)
+
+	var rhsG1 bls.G1Point
+	bls.SubG1(&rhsG1, aggCommit, aggPolyG1)
+	bls.AddG1(&rhsG1, &rhsG1, offsetG1)
+	return &rhsG1, nil
+}
+
+// UniversalVerify implements batch verification on a set of chunks given the same chunk dimension (chunkLen, numChunk).
+// The details is given in Ethereum Research post whose authors are George Kadianakis, Ansgar Dietrichs, Dankrad Feist
+// https://ethresear.ch/t/a-universal-verification-equation-for-data-availability-sampling/13240
+//
+// m is number of blob, samples is a list of chunks
+//
+// The order of samples do not matter.
+// Each sample need not have unique row, it is possible that multiple chunks of the same blob are validated altogether
```

Comment:
`it is possible` -> `it is always`? The chunks from the same blob always share the same encoding param.

Reply:
but a operator might only be assigned 1 chunk in one blob

Reply:
The context here is that all chunks assigned to this operator (will be always validated together)